xla/service/cpu/onednn_matmul_rewriter.cc

```diff
@@ -91,6 +104,80 @@ inline void GetBF16Bias(HloInstruction* dot, HloInstruction** old_bias,
 
+inline auto BcastConstScalar(HloInstruction** instr, double value) {
+  return m::Broadcast(instr, m::ConstantScalar(value));
+}
+
+inline auto BcastConstScalar(double value) {
+  return BcastConstScalar(nullptr, value);
+}
+
+auto ConstScalarNear(double value) {
+  return m::ConstantScalar().WithPredicate(
+      [expected = value](const HloInstruction* instr) {
+        // Not a very robust floating-point comparison, but good enough for our
+        // purposes.
+        std::optional<double> actual =
+            static_cast<const HloConstantInstruction*>(instr)
+                ->literal()
+                .GetAsDouble({});
+        if (!actual.has_value()) return false;
+        double epsilon;
+        switch (instr->shape().element_type()) {
+          case F16:
+            epsilon = 128 * std::numeric_limits<Eigen::half>::epsilon();
+            break;
+          case BF16:
+            epsilon = 128 * std::numeric_limits<bfloat16>::epsilon();
+            break;
+          case F32:
+            epsilon = 128 * std::numeric_limits<float>::epsilon();
+            break;
+          case F64:
+            epsilon = 128 * std::numeric_limits<double>::epsilon();
+            break;
+          default:
+            return false;
+        }
+        return abs(*actual - expected) < (abs(*actual + expected) * epsilon);
+      });
+}
+
+inline auto BcastConstScalarNear(double value) {
+  return m::Broadcast(ConstScalarNear(value));
+}
+
+auto GELUActivation(HloInstruction* instr, HloInstruction** src) {
+  // Attempt to match GELU_TANH activation
+  // (https://arxiv.org/abs/1606.08415), where:
+  // gelu_tanh(x) = x * cdf(x)
+  // cdf(x) = 0.5 * (1 + tanh(sqrt(2 / pi) * (x + 0.044715 * x**3))
+  HloInstruction* errf;
+  return Match(instr, m::MultiplyAnyOrder(
+                          m::Op(src),
+                          m::MultiplyAnyOrder(
+                              BcastConstScalar(0.5),
+                              m::AddAnyOrder(BcastConstScalar(1.0),
+                                             m::Op(&errf).WithOneUser())))) &&
+         Match(errf,
+               m::Tanh(m::MultiplyAnyOrder(
+                           BcastConstScalarNear(sqrt(M_2_PI)),
+                           m::AddAnyOrder(
+                               m::Op().Is(*src),
+                               m::MultiplyAnyOrder(
+                                   BcastConstScalarNear(0.044715),
+                                   m::MultiplyAnyOrder(
+                                       m::Op().Is(*src),
+                                       m::MultiplyAnyOrder(m::Op().Is(*src),
+                                                           m::Op().Is(*src))
+                                           .WithOneUser())
+                                       .WithOneUser())
+                                   .WithOneUser())
+                               .WithOneUser())
+                           .WithOneUser())
+                   .WithOneUser());
```

Comment:
Since this pattern is more specific, would the check be faster if we try to match `errf` first?

Reply:
Good note.
Unfortunately, as I understand, to start matching from `errf`, we have to handle `Tanh` and then track operands back up to CDF then up to MatMul.
I am afraid, such path can be too complex and hard to maintain.

The similar GELU pattern matching is implemented for GPU.
IMHO, would be best to implement common code to match ML operations patterns (e.g. GELU, Sofmax, etc..) and reuse them in different backends.

Reply:
Sounds good. Thank you for the explanation!